src/BaGetter.Core/Indexing/PackageDeletionService.cs

```diff
@@ -90,21 +91,106 @@ private async Task<bool> TryHardDeletePackageAsync(string id, NuGetVersion versi
-        var versions = await _packages.FindAsync(package.Id, includeUnlisted: true, cancellationToken);
-        if (versions is null || versions.Count <= maxPackages) return 0;
+        var packages = await _packages.FindAsync(package.Id, includeUnlisted: true, cancellationToken);
+        if (packages is null || packages.Count <= maxMajor) return 0;
+
+        var goodVersions = new HashSet<NuGetVersion>();
+
+        if (maxMajor.HasValue)
+        {
+            goodVersions = GetValidVersions(packages.Select(t => t.Version), v => 0, v => v.Major, (int)maxMajor).ToHashSet();
+        }
+        else
+        {
+            goodVersions = packages.Select(p => p.Version).ToHashSet();
+        }
+        
+        if (maxMinor.HasValue)
+        {
+            goodVersions.IntersectWith(GetValidVersions(goodVersions, v => (v.Major), v => v.Minor, (int)maxMinor));
+        }
+        
+        if (maxPatch.HasValue)
+        {
+            goodVersions.IntersectWith(GetValidVersions(goodVersions, v => (v.Major, v.Minor), v => v.Patch, (int)maxPatch));
+        }
+
+        if (maxPrerelease.HasValue)
+        {
+            // this assume we have something like 1.1.1-alpha.1 - alpha is the release type
+            var preReleases = packages.Select(p => p.Version).Where(p => p.IsPrerelease).ToList();
+            // this will give us 'alpha' or 'beta' etc
+            var prereleaseTypes = preReleases
+                .Select(v => v.ReleaseLabels?.FirstOrDefault())
+                .Where(lb => lb is not null)
+                .Distinct();
+
+            var allPreReleaseValidVersions = new HashSet<NuGetVersion>();
+            foreach (var preReleaseType in prereleaseTypes)
+            {
+                var preReleaseVersions = preReleases.Where(p => p.ReleaseLabels!.FirstOrDefault() == preReleaseType
+                        && GetPreReleaseBuild(p) is not null).ToList();
+
+                
+                allPreReleaseValidVersions.UnionWith
+                    (GetValidVersions(preReleaseVersions,
+                        v => (v.Major, v.Minor, v.Patch), v => GetPreReleaseBuild(v).Value, (int)maxPrerelease));
+
+            }
+            goodVersions.IntersectWith(allPreReleaseValidVersions);
+        }
+
         // sort by version and take everything except the last maxPackages
-        var versionsToDelete = versions
-            .OrderByDescending(p => p.Version)
-            .Skip((int)maxPackages)
-            .ToList();
+        var versionsToDelete = packages.Where(p => !goodVersions.Contains(p.Version)).ToList();
+
         var deleted = 0;
         foreach (var version in versionsToDelete)
         {
             if (await TryHardDeletePackageAsync(package.Id, version.Version, cancellationToken)) deleted++;
         }
         return deleted;
     }
+
+    /// <summary>
+    /// If we have 1.1.1-alpha.1 , this will return 1
+    /// or null if not valid
+    /// </summary>
+    /// <returns></returns>
```

Comment:
```suggestion
    /// <summary>
    /// Tries to get the version number of a pre-release build.<br/>
    /// If we have 1.1.1-alpha.1 , this will return 1 or <c>null</c> if not valid.
    /// </summary>
    /// <returns>The version as <c>int</c> or <c>null</c> if not found.</returns>
```